doofinder-for-woocommerce/includes/class-landing.php

```diff
@@ -352,6 +309,76 @@ public function get_disabled_html()
         return $html;
     }
 
+    /**
+     * Determine the desired language based on the provided hashid and a list of languages.
+     *
+     * @param array $languages An array of available languages with their properties.
+     * @param string $hashid The hashid associated with the desired language.
+     *
+     * @return string The code of the desired language, or an empty string if not found.
+     */
+    private function get_desired_language($languages, $hashid) {
+        $desired_lang = '';
+        foreach ($languages as $language) {
+            $hashid_by_lang = Settings::get_search_engine_hash($language['code']);
+            if($hashid_by_lang === $hashid) {
+                $desired_lang = $language['code'];
+                break;
+            }
+        }
+
+        return $desired_lang;
+    }
+
+    /**
+     * Format the URL by appending the slug to the home URL while considering different URL formats.
+     *
+     * @param string $home_url The home URL for the desired language.
+     * @param string $slug The slug to be appended to the URL.
+     *
+     * @return string The formatted URL with the slug included.
+     */
+    private function formated_url($home_url, $slug) {
+        $slug = "df/{$slug}";
+        if(strpos($home_url, '?lang=') !== false) {
+            $formated_url = str_replace('?lang=', "{$slug}/?lang=", $home_url);
+        } else {
+            $formated_url = "{$home_url}{$slug}";
+        }
+        return $formated_url;
+    }
+
+    /**
+     * Redirect to a given formatted URL.
+     *
+     * @param string $formated_url The URL to which the redirection will occur.
+     */
+    private function redirect($formated_url) {
+        header("Location: $formated_url");
+        exit;
+    }
+
+    private function request_landing_info($hashid, $slug) {
+        $lang_cache = self::DF_LANDING_CACHE . "_" . $this->current_language;
```

Comment:
You're using this concatenation on two different locations. I think it can be unified into a independent function, but it might not be needed (just a suggestion :smile: )